library/cartridge_edit_topology.py

```diff
@@ -320,20 +322,113 @@ def get_edit_failover_priority_params(replicasets, cluster_replicasets, cluster_
             continue
 
         if cluster_replicaset['instances'][:len(failover_priority)] != failover_priority:
-            failover_priority_uuids = [
-                cluster_instances[instance_name]['uuid'] for instance_name in failover_priority
-                if instance_name in cluster_instances  # false if instance is expelled
-            ]
+            failover_priority_uuids = []
+            for instance_name in failover_priority:
+                if instance_name not in cluster_instances or not cluster_instances[instance_name].get('uuid'):
+                    return None, "Instance %s from %s failover_priority isn't joined to cluster" % (
+                        instance_name, alias
+                    )
+                failover_priority_uuids.append(cluster_instances[instance_name]['uuid'])
 
             edit_replicasets_params.append({
                 'uuid': cluster_replicaset['uuid'],
                 'failover_priority': failover_priority_uuids,
             })
 
+    return edit_replicasets_params, None
+
+
+def add_edit_server_param_if_required(edit_server_params, instance_params, cluster_instance, param_name):
+    if instance_params.get(param_name) is None:
+        return
+
+    if cluster_instance is not None:
+        if instance_params.get(param_name) == cluster_instance.get(param_name):
+            return
+
+    edit_server_params[param_name] = instance_params.get(param_name)
+
+
+def get_edit_server_params(instance_name, instance_params, cluster_instances):
+    if instance_name not in cluster_instances:
+        if instance_params.get('expelled') is True:
+            return None, None
+        return None, "Instance %s isn't found in cluster" % instance_name
+
+    cluster_instance = cluster_instances[instance_name]
+
+    if not cluster_instance.get('uuid'):  # uuid is '' for unjoined instances
+        return None, None
+
+    edit_server_params = {
+        'uuid': cluster_instance.get('uuid'),
+    }
+
+    if instance_params.get('expelled') is True:
+        edit_server_params['expelled'] = True
+    else:
+        add_edit_server_param_if_required(
+            edit_server_params, instance_params, cluster_instance, 'zone'
+        )
+
+    if len(edit_server_params) == 1:
+        # all instance parameters are the same as configured
+        return None, None
+
+    return edit_server_params, None
+
+
+def get_edit_servers_params(instances, cluster_instances):
+    edit_servers_params = []
+    for instance_name, instance_params in instances.items():
+        edit_server_params, err = get_edit_server_params(instance_name, instance_params, cluster_instances)
+        if err is not None:
+            return None, "Failed to get edit topology params for instance %s: %s" % (instance_name, err)
```

Comment:
Isn't error message too long?
`Failed to get edit topology params for instance NAME: Instance NAME isn't found in cluster`

Reply:
Why not?